mmds/src/data_store.rs

```diff
@@ -35,15 +35,43 @@ impl Mmds {
         self.is_initialized
     }
 
-    pub fn put_data(&mut self, data: Value) {
-        // TODO: we should add a data validator and only accept Strings, arrays & dictionaries
-        // https://github.com/firecracker-microvm/firecracker/issues/401
+    /// This method validates the data from a PATCH or PUT request and returns true
+    /// if the data only contains Strings, arrays and dictionaries
+    pub fn is_data_valid(&self, data: &Value) -> bool {
+        if let Some(map) = data.as_object() {
+            for key in map.keys() {
+                if !self.is_data_valid(&map[key]) {
+                    return false;
+                }
+            }
+            true
+        } else if let Some(array) = data.as_array() {
+            for obj in array {
+                if !self.is_data_valid(&obj) {
+                    return false;
+                }
+            }
+            true
+        } else {
+            data.is_string()
+        }
+    }
+
+    pub fn put_data(&mut self, data: Value) -> Result<(), Error> {
+        if !self.is_data_valid(&data) {
+            return Err(Error::UnsupportedValueType);
+        }
         self.data_store = data;
         self.is_initialized = true;
+        Ok(())
     }
 
-    pub fn patch_data(&mut self, patch_data: Value) {
+    pub fn patch_data(&mut self, patch_data: Value) -> Result<(), Error> {
```

Comment:
Since this function now can return an Error, we should also have a unit test for the error case. But as it might be a lot of duplicated code since we are also testing `is_data_valid`, we can have one test for `put_data` and one test for `patch_data` and adapt the code for `is_data_valid`.

My suggestion would be as follows:
- put_data with valid data (all strings) -> check `put_data().is_ok()`
- patch_data with valid data (all strings) -> check `patch_data().is_ok()`
- Error Case 1: patch_data with invalid data (one bool or one number) -> check `patch_data() == Err(Error:UnsupportedValueType)` 
- Error Case 2: put_data with invalid data (one bool or one string) -> check `put_data() == Err(Error:UnsupportedValueType)`.

For Error Case 1 you can use this JSON (because the invalid value is in a Map):
```rust
let data = r#"{
            "name": {
                "first": "John",
                "second": "Doe"
            },
            "age": 43
        }"#;
```

For Error Case 2 you can use this JSON (because the invalid value is in an array):
```rust
let data = r#"{
            "name": {
                "first": "John",
                "second": true
            }
        }"#;
```

With these test cases we should have everything cover. For simplicity you can have them in one test something like `test_update_data_store()`.